skimage/util/_backends.py

```diff
@@ -0,0 +1,146 @@
+    else:
+        selected_entry_points = all_entry_points.get(group, ())
+    return selected_entry_points
+
+
+def dispatching_disabled():
+    """Determine if dispatching has been disabled by the user"""
+    no_dispatching = os.environ.get("SKIMAGE_NO_DISPATCHING", False)
+    if no_dispatching == "1":
+        return True
+    else:
+        return False
+
+
+def public_api_name(func):
+    """Get the name of the public module for a scikit-image function.
+
+    This computes the name of the public submodule in which the function can
+    be found.
+    """
+    full_name = func.__module__
+    # This relies on the fact that scikit-image does not use
+    # sub-submodules in its public API, except in one case.
+    # This means that public name can be atmost `skimage.foobar`
+    # for everything else
+    if full_name.startswith("skimage.filters.rank"):
+        public_name = "skimage.filters.rank"
+    else:
+        public_name = ".".join(full_name.split(".")[:2])
+
+    # It would be nice to sanity check things by doing something like the
+    # following. However we can't because this code is executed while the
+    # module is being imported, which means this would create a circular
+    # import
+    # mod = importlib.import_module(public_name)
+    # assert getattr(mod, func.__name__) is func
+
+    return public_name
+
+
+@cache
+def all_backends():
+    """List all installed backends and information about them"""
+    backends = {}
+    backends_ = _entry_points(group="skimage_backends")
+    backend_infos = _entry_points(group="skimage_backend_infos")
+
+    for backend in backends_:
+        backends[backend.name] = {"implementation": backend}
+        try:
+            info = backend_infos[backend.name]
+            # Double () to load and then call the backend information function
+            backends[backend.name]["info"] = info.load()()
+        except KeyError:
+            pass
+
+    return backends
+
+
+def dispatchable(func):
+    """Mark a function as dispatchable.
+
+    When a decorated function is called the installed backends are
+    searched for an implementation. If no backend implements the function
+    then the scikit-image implementation is used.
+    """
+    func_name = func.__name__
+    func_module = public_api_name(func)
+
+    # If no backends are installed at all or dispatching is disabled,
+    # return the original function. This way people who don't care about it
+    # don't see anything related to dispatching
+    if dispatching_disabled() or not all_backends():
+        return func
+
+    @functools.wraps(func)
+    def wrapper(*args, **kwargs):
+        # Backends are tried in alphabetical order, this makes things
+        # predictable and stable across runs. Might need a better solution
+        # when it becomes common that users have more than one backend
+        # that would accept a call.
+        for name in sorted(all_backends()):
```

Comment:
> TODO: Taking backend from the user and dispatching to that backend instead of sorting the backends and dispatching to the first backend that has the function's implemented.

Maybe something to discuss in a new issue? I don't think a user should be able to select a backend when calling a function via something like `skimage.foo.bar(..., backend="use_this_backend")`. Maybe a context manager is the way to enable/disable backends temporarily.



> And do we want dispatching based on just array types or do we want to allow multiple backends supporting the same array type but having different implementations?

I think dispatching only on types is too limiting. In this PR a backend can be chosen using the maximum information available (the function name and the values passed to the function). If a backend only wants to handle a subset of array types, it can do that. Similarly if a backend only wants to handle very large input arrays of any type, it can do that.

I'm not sure I understand the bit about the qualname :-/

Reply:
> Maybe a context manager is the way to enable/disable backends temporarily.

I am not sure if having a context manager would be useful to the scikit-image users or not. We can instead have an environment variable that stores the backend's name and I think that might be better way here because then the users wouldn't have to change their skimage code alot or write it all in a context manager. 

In NetworkX we have the `NETWORKX_BACKEND_PRIORITY` env variable. It can be either just one backend_name or a list of backend_names. And in networkx you can also dispatch using a context manager but the context manager there was introduced because we wanted to offer backend configurations. But, I'm not sure if that's something that would be very useful in the scikit-image and its backends ecosystem. (ref- https://networkx.org/documentation/latest/reference/backends.html). 

> I'm not sure I understand the bit about the qualname :-/

That's something that Erik told me during the euroscipy sprints, on how we might be able to do type-based dispatching with arrays. But, I don't think we want to do that in scikit-image. Do you know how does(or if) sklearn or scipy uses type-based dispatching with arrays?